nanocoap: Add blockwise transfer block2 support

bergzand opened the pull request:
This PR adds block2 support for nanocoap. This facilitates larger transfers such as a large /.well-know/core payload.

I've tried to make it as easy as possible for a handler to support blockwise transfer. As an example I rewrote the /.well-known/core handler with block2 support.

I've added two extra fields to the `coap_pkt_t` struct, but with #18 merged, it should be possible to remove these fields.

haukepetersen committed:
nanocoap: added function for finding options

bergzand committed:
nanocoap: Add header definitions from rfc 7959

bergzand committed:
nanocoap: Add block2 support

bergzand committed:
nanocoap: Minor style fixes

bergzand commented:
I need a small bit of advise before continuing with the size2 option. The problem with both the block2 and the size2 option is that the value needs to be adjusted after the payload is inserted. The block2 header needs adjustment if the last block is reached and the size2 requires the full size of the payload.

This problem can be resolved with #18, finding these options again after the payload is written and adjusting them is possible with this. Furthermore, the two additional fields in the `coap_pkt_t` struct can be removed again.

So with this in mind, should I continue with a rebase on #18 and assume that it will be merged eventually?

kaspar030 commented:
> So with this in mind, should I continue with a rebase on #18 and assume that it will be merged eventually?

Yes!

kaspar030 commented:
Thanks for tackling this.
Please rebase (probably to #18), and prefix the commits with "nanocoap:".

bergzand committed:
nanocoap: small opimization

haukepetersen committed:
fixup! nanocoap: fixed things

haukepetersen committed:
fixup! pass payload position directly to find_option()

bergzand committed:
nanocoap: use find_option to retrieve block2 header

bergzand commented:
Rebased and adjusted for `coap_find_option`.
I've split the `coap_put_option_block2` into an extra `coap_finish_option_block2` to reduce some checks and improve readability.

bergzand committed:
nanocoap: Add public functions to header file

bergzand committed:
fixup! nanocoap: add header existence check

bergzand commented:
Still contains a few bugs with larger resources where the option length is longer than 1.

bergzand committed:
Fixup! nanocoap: fix integer overflows, windowing math

bergzand commented:
Tested some more with really huge transfers, everything seems to work now. If somebody could check if [r421](https://github.com/kaspar030/sock/pull/21/files#diff-dabba578e61a26201ee43e794251ae44R421) is endianess proof, I'd be happy.

bergzand commented:
Closing this now that it is merged in RIOT.
